### PR TITLE
Allocate semantic checks (second and last part)

### PR DESCRIPTION
**Pull request by jeanPerier**
Implement semantic checks and related tests for constraints:
C937, C938, C939, C940, C941, C942, C945 (second part),
C946, C947, C948, C949 and C950.

### CODE REVIEW

File: lib/semantics/tools.cc [hunk 1]
```diff
@@ -308,4 +308,93 @@ const Symbol *FindFunctionResult(const Symbol &symbol) {
...
+}
+
+const Symbol *HasCoarrayUltimateComponent(
+    const DerivedTypeSpec &derivedTypeSpec) {
+  const Symbol &symbol{derivedTypeSpec.typeSymbol()};
+  // TODO is it guaranteed that derived type symbol have a scope and is it the
+  // right scope to look into?
+  CHECK(symbol.scope());
+  for (const Symbol *componentSymbol :
+      symbol.get<DerivedTypeDetails>().OrderComponents(*symbol.scope())) {
```
Comment (jeanPerier) [inline lib/semantics/tools.cc]: @klausler or @tskeith , is this the right scope to use here?

File: lib/semantics/check-allocate.cc [hunk 2]
```diff
@@ -37,10 +40,54 @@ struct AllocateCheckerInfo {
...
+    : allocateInfo_{info}, allocateObject_{std::get<parser::AllocateObject>(
+                               alloc.t)},
+      name_{parser::GetLastName(allocateObject_)},
       type_{name_.symbol ? name_.symbol->GetType() : nullptr},
-      isSubobject_{std::holds_alternative<parser::StructureComponent>(obj.u)} {
+      allocateShapeSpecRank_{ShapeSpecRank(alloc)},
+      rank_{name_.symbol ? name_.symbol->Rank() : 0},
+      allocateCoarraySpecRank_{CoarraySpecRank(alloc)},
+      corank_{name_.symbol ? name_.symbol->Corank() : 0} {
+    GatherAllocationBasicInfo();
```
Comment (klausler) [inline lib/semantics/check-allocate.cc]: Constructors shouldn't do a lot of work or throw exceptions.  Please consider calling `GatherAllocationBasicInfo()` after the object has been constructed.

File: lib/semantics/check-allocate.cc [hunk 3]
```diff
@@ -37,10 +40,54 @@ struct AllocateCheckerInfo {
...
+  }
+
+  bool RunChecks(SemanticsContext &context);
+
+private:
+  AllocateCheckerInfo &allocateInfo_;
```
Comment (klausler) [inline lib/semantics/check-allocate.cc]: It's a small point, but we generally put data members last in a class definition, after any private member functions.

File: test/semantics/allocate11.f90 [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  type D
+    type(A), pointer :: potential
+  end type
+
+
+
+  class(*), allocatable :: var
+  ! unlimited polymorphic is the ONLY way to get an allocatable/pointer 'var' that can be
+  ! allocated with a type-spec T that has coarray ultimate component.
+  ! Rational:
```
Comment (klausler) [inline test/semantics/allocate11.f90]: Spelling: "Rationale"

File: lib/semantics/check-allocate.cc (hunk 2 above)
Comment (jeanPerier) [reply lib/semantics/check-allocate.cc]: Ok, I moved `GatherAllocationBasicInfo()` into `RunChecks(SemanticsContext &context)`.

File: lib/semantics/check-allocate.cc (hunk 3 above)
Comment (jeanPerier) [reply lib/semantics/check-allocate.cc]: Fixed

File: test/semantics/allocate11.f90 (hunk 4 above)
Comment (jeanPerier) [reply test/semantics/allocate11.f90]: Fixed. Also rephrased part of the comment that were wrong.

**Pull request merged**